core/safemath/safe_math.go

```diff
@@ -0,0 +1,170 @@
+	//
+	// As a positive overflow example, take int8(66) * 2.
+	// The result as a uint16 will be 132, or in binary: 0000 0000 1000 0100.
+	// Interpreted as an int8 the high part is cut off and so the sign bit (most significant bit) is set,
+	// but we expected the result to be positive, so we detect this as an overflow.
+	//
+	// *Negative Overflow*
+	//
+	// As a negative overflow example, take int8(-66) * 2.
+	// The result as a uint16 will 132, or in binary 0000 0000 1000 0100, since we calculate 66*2 without the sign.
+	// Interpreted as an int8 the high part is cut off and the result multiplied with -1, to account for the
+	// expected sign of the result, so we get 0111 1100.
+	// The sign bit (most significant bit) is unset, but we expected the result to be negative,
+	// so we detect this as an overflow.
+	//
+	// *Overflowing 8-bit space*
+	//
+	// As a larger negative overflow example using 8-bit integers, take int8(-128) * 32.
+	// The multiplication would take place using 16-bits, so the result as a uint16 will be
+	// binary 0001 0000 0000 0000, since we calculate 128*32 without the sign.
+	// Interpreted as an int8 and multiplied with -1, to account for the expected sign of the result, we simply get
+	// 0000 0000, since the least significant byte of the result is 0.
+	// However, since the most significant byte of the result is non-zero, we detect the overflow.
+
+	xNegative := x < 0
+	yNegative := y < 0
+	xPositive := x > 0
+	yPositive := y > 0
+	resultIsPositive := true
+	var resultSign int64 = 1
+
+	// Store the sign of the resulting computation and negate any negative integers,
+	// so we can correctly convert to uint64.
+	if xNegative {
+		if yPositive {
+			resultIsPositive = false
+			resultSign = -1
+		}
+		x = -x
+	}
+
+	if yNegative {
+		if xPositive {
+			resultIsPositive = false
+			resultSign = -1
+		}
+		y = -y
+	}
+
+	// Execute the multiplication in 128-bit space using unsigned integers.
+	hi, lo := bits.Mul64(uint64(x), uint64(y))
+	// Interpret the lo result as an int64, then correct for the expected sign.
+	loSigned := int64(lo) * resultSign
+
+	// If the computation overflowed a uint64, it would also overflow an int64.
+	if hi != 0 {
+		return 0, ierrors.Wrapf(ErrIntegerOverflow, "%d and %d", x, y)
+	}
+
+	// Extract the most significant bit, signaling if the number is negative (1) or not (0).
+	signBitSet := ((loSigned >> 63) & 1) == 1
+
+	// If the result is expected to be positive but the sign bit is set, it's an overflow.
+	if resultIsPositive && signBitSet {
+		return 0, ierrors.Wrapf(ErrIntegerOverflow, "%d and %d", x, y)
+	}
+
+	// If the result is expected to be negative but the sign bit is not set, it's an underflow.
+	if !resultIsPositive && !signBitSet {
+		return 0, ierrors.Wrapf(ErrIntegerOverflow, "%d and %d", x, y)
+	}
+
+	return loSigned, nil
+}
+
+// Returns x * y or an error if that computation would under- or overflow.
+func SafeMul[T Integer](x T, y T) (T, error) {
+	result := x * y
```

Comment:
You could check if y or x are zero directly at the start and return 0 as an result.
Then you can remove one condition in the next if and it should be faster in case one of the vars is 0.

Reply:
Maybe you should check for 0 for every "Mul" function.

Reply:
> You could check if y or x are zero directly at the start and return 0 as an result.

Right now there are two conditions checked, and if we replace the later `x != 0` check with two checks `x == 0 || y == 0` we have three conditions checked in total. Wouldn't that be slower in the general case when x and y are non-zero? Or in other words, I'm not sure if checking for 0 upfront to make the special case where `x` or `y` are zero a tiny bit faster at the cost of making all other cases a tiny bit slower.

Reply:
I would do it anyway. The performance improvement for the case where one of the factors is 0 is huge. And maybe with the branch prediction of modern CPUs you don't loose performance at all for that check.